cc-wrapper: expand response files

corngood opened the pull request:
###### Motivation for this change

Some arguments were not handled by the gcc wrapper when passed in a response file.  When given a response file, gcc will in turn pass a response file to ld containing all arguments, not just the ones from the original reponse file.

I've tested it with quoted strings, escaped quotes, escaped spaces, newlines, etc.  It should interpret things the same as on the command line, but let me know if you can think of anything else I should test.

TODO:
- pack large argument lists in a temporary reponse file?
  - platform specific (e.g. cygwin)?
  - only when a response file is encountered?
- tests
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Fixes #11762

corngood committed:
cc-wrapper: expand response files

Fixes #11762

corngood commented:
Any advice on testing this properly without building _everything_ on my machine?

copumpkin commented:
cc @shlevy too, who's been interested in this recently.

shlevy commented:
God I hate how messy this wrapper is. But yeah, for now this seems like the right approach.

To test it, I think the best bet is to just copy the whole cc-wrapper stuff to a new dir and write a new wrapCCWith thing using that copy, and use that to build a new stdenv to build some package or other.

corngood commented:
@shlevy Sorry, I didn't really explain that very well.  I've tested it locally on some subset of packages.  I'm just wondering if there's any thing I can do to make sure this doesn't break some obscure package, without building everything on my machine.  The Travis CI doesn't seem to be working properly.

shlevy commented:
Oh, this PR should be opened against the `staging` branch, mass rebuilds are what it's for.

corngood commented:
Tagged WIP, because I found a problem building bison.  failed tests:

```
 81: Output file name: @@                            FAILED (output.at:229)
 82: Output file name: @{                            FAILED (output.at:230)
 83: Output file name: @}                            FAILED (output.at:231)
```

corngood commented:
Rebased on staging, and fixed the problem shown in the bison test.

gcc interprets arguments starting with @ as a response file only if the file exists, and otherwise treats them as regular arguments.
